ndarray/src/main/java/org/tensorflow/ndarray/index/Slice.java

```diff
@@ -0,0 +1,111 @@
+/*
+  Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow.ndarray.index;
+
+import java.util.StringJoiner;
+import org.tensorflow.ndarray.impl.dimension.Dimension;
+
+final class Slice implements TensorIndex {
+
+  private final Long start;
+  private final Long end;
+  private final long stride;
+
+  private long start(Dimension dim){
+    if(start == null){
+      if(stride > 0){
+        return 0;
+      } else {
+        return dim.numElements() - 1; // it's inclusive
+      }
+    } else if(start < 0){
+      return dim.numElements() + start;
+    } else {
+      return start;
+    }
```

Comment:
Nit: no need of `else` clauses when you return unconditionally in your previous `if` blocks, I find it improves readability

Also not required, most of the code in TF Java follow a reversed order of the elements of a class, where elements with a higher visibility (public) are at the top of the class while those at the bottom has the lower one.